### DISCUSSION TITLE
Full stack web development using frontend GHCJS libraries

### DISCUSSION THREAD

**Discussion by newton-migosi** [srid/haskell-flake#184] [Q&A]
Is there a simple example of using `haskell-flake` for full stack web apps? 

I'm setting up a project that uses [`miso`](https://github.com/dmjio/miso) as the frontend web framework. `miso` is nixified but is not a flake.
- how do we provision both GHC for the backend and GHCJS for the frontend?
- how do we set up `cabal.project` and the `.cabal` files for a full stack project in a way that `haskell-flake` can detect?

I've started by creating a package set with ghcjs following https://zero-to-flakes.com/haskell-flake/dependency and https://zero-to-flakes.com/haskell-flake/package-set
```nix
haskellProjects.miso-ghc = {
          defaults.packages = {};  # No local packages
          devShell.enable = false;

          basePackages = pkgs.haskell.packages.ghcjs86;

          packages = {
            miso = inputs.miso;
            servant-jsaddle = inputs.servant-jsaddle;
          };

          settings = {
            servant-jsaddle = {
              check = false;
            };
          };
 };
```
using the inputs
```nix
inputs = {
  miso = { url = "github:dmjio/miso"; flake = false; };
  servant-jsaddle = { url = "github:haskell-servant/servant-jsaddle/master"; flake = false; };
};
```

**Comment by newton-migosi** [comment]
my current directory structure looks like (abridged)
```
.
├── backend
│   ├── app
│   │   └── Main.hs
│   ├── fourmolu.yaml
│   ├── hie.yaml
│   ├── local-library.cabal
│   ├── src
│   │   └── LocalLibrary
│   │       ├── API.hs
│   │       └── Greetings
│   │           ├── API.hs
│   │           └── SampleData.hs
│   ├── test
│   │   └── Main.hs
├── flake.lock
├── flake.nix
├── frontend
│   ├── app
│   │   └── Main.hs
│   ├── cabal.project
│   ├── CHANGELOG.md
│   ├── hie.yaml
│   ├── LICENSE
│   ├── local-library-frontend.cabal
│   ├── src
│   │   └── MyLib.hs
│   └── test
│       └── Main.hs
├── Jenkinsfile
├── LICENSE
├── README.md
└── shell.nix


```

**Comment by srid** [comment]
> `miso` is nixified but is not a flake.

This is what you want to figure out first; how do you develop a miso project using a regular flake (using just nixpkgs)? Going from there to using it in haskell-flake is a straightforward next step.

**Reply by srid** [reply]
Last I checked, miso recommended using reflex-platform which is whole another complex beast. Not sure if things have improved on ghcjs ecosystem since then ...

**Reply by newton-migosi** [reply]
I'm not familiar enough with nix to start a flake project based on miso.. is there another ghcjs library that already supports nix flakes?

**Reply by srid** [reply]
Not sure, you can try asking in: https://app.element.io/#/room/#haskell:nixos.org or the Miso channel.

You are looking for something that uses `callCabal2nix` and `shellFor`. Once you have that, it is straightforward to port that to haskell-flake.

**Reply by srid** [reply]
> You are looking for something that uses `callCabal2nix` and `shellFor`

https://zero-to-flakes.com/nixpkgs-haskell
